### ISSUE TITLE
Deploying my own crnn model using triton, the output has the wrong shape

### ISSUE THREAD

**Issue by llustc** [triton-inference-server/server#3989]
### triton version
- 21.09

### config.pbtxt
name: "crnn"
platform: "pytorch_libtorch"
backend: "pytorch",
max_batch_size: 16
input [
  {
    name: "input__0"
    data_type: TYPE_FP32
    format: FORMAT_NCHW
    dims: [ 1, 32, 320 ]
  }
]

output [
  {
    name: "output__0"
    data_type: TYPE_FP32
    dims: [81, -1,  7219]
  }
]

model_warmup: [
  {
    name: "random_input__0",
    batch_size: 1,
    inputs: {
      key: "input__0"
      value:  {
        data_type: TYPE_FP32,
        dims: [ 1, 32, 320 ],
        random_data: true
      }
    }
  }
]

### Problem Description
-Using pytorch to test the model output shape is (81, 1, 7219)
-Using triton to test the model output shape is (1, 1, 7219)
-No matter what the shape of the output is set to in the configuration file, the output of triton is unchanged. Of course, if the output shape is not set, the triton will report an error when it starts.

### logs 
-pytorch
>>> import torch
>>> model = torch.jit.load("model.pt")
>>> input = torch.rand(1, 1, 32, 320)
>>> model(input).shape
torch.Size([81, 1, 7219])

-triton
original inputs:
[0x0x7fab2f642cf8] input: input__0, type: FP32, original shape: [1,1,32,320], batch + shape: [1,1,32,320], shape: [1,32,320]
override inputs:
inputs:
[0x0x7fab2f642cf8] input: input__0, type: FP32, original shape: [1,1,32,320], batch + shape: [1,1,32,320], shape: [1,32,320]
original requested outputs:
requested outputs:
output__0

I0228 08:38:46.597296 28494 triton_model_instance.cc:667] Starting backend thread for crnn_0 at nice 0 on device 0...
I0228 08:38:46.597389 28494 triton_model_instance.cc:531] model 'crnn' instance crnn_0 is running warmup sample 'random_input__0'
I0228 08:38:46.597442 28494 libtorch.cc:1185] model crnn, instance crnn_0, executing 1 requests
I0228 08:38:46.597458 28494 libtorch.cc:568] TRITONBACKEND_ModelExecute: Running crnn_0 with 1 requests
I0228 08:38:46.597534 28494 pinned_memory_manager.cc:161] pinned memory allocation: size 40960, addr 0x7fabf200a0a0
I0228 08:38:48.745245 28494 infer_response.cc:165] add response output: output: output__0, type: FP32, shape: [1,1,7219]

What is the possible reason for this problem，and how do i fix it? 

### my model struct
RecursiveScriptModule(
  original_name=CRNN
  (cnn): RecursiveScriptModule(
    original_name=Sequential
    (conv0): RecursiveScriptModule(original_name=Conv2d)
    (relu0): RecursiveScriptModule(original_name=ReLU)
    (pooling0): RecursiveScriptModule(original_name=MaxPool2d)
    (conv1): RecursiveScriptModule(original_name=Conv2d)
    (relu1): RecursiveScriptModule(original_name=ReLU)
    (pooling1): RecursiveScriptModule(original_name=MaxPool2d)
    (conv2): RecursiveScriptModule(original_name=Conv2d)
    (batchnorm2): RecursiveScriptModule(original_name=BatchNorm2d)
    (relu2): RecursiveScriptModule(original_name=ReLU)
    (conv3): RecursiveScriptModule(original_name=Conv2d)
    (relu3): RecursiveScriptModule(original_name=ReLU)
    (pooling2): RecursiveScriptModule(original_name=MaxPool2d)
    (conv4): RecursiveScriptModule(original_name=Conv2d)
    (batchnorm4): RecursiveScriptModule(original_name=BatchNorm2d)
    (relu4): RecursiveScriptModule(original_name=ReLU)
    (conv5): RecursiveScriptModule(original_name=Conv2d)
    (relu5): RecursiveScriptModule(original_name=ReLU)
    (pooling3): RecursiveScriptModule(original_name=MaxPool2d)
    (conv6): RecursiveScriptModule(original_name=Conv2d)
    (batchnorm6): RecursiveScriptModule(original_name=BatchNorm2d)
    (relu6): RecursiveScriptModule(original_name=ReLU)
  )
  (rnn): RecursiveScriptModule(
    original_name=Sequential
    (0): RecursiveScriptModule(
      original_name=BidirectionalLSTM
      (rnn): RecursiveScriptModule(original_name=LSTM)
      (embedding): RecursiveScriptModule(original_name=Linear)
    )
    (1): RecursiveScriptModule(
      original_name=BidirectionalLSTM
      (rnn): RecursiveScriptModule(original_name=LSTM)
      (embedding): RecursiveScriptModule(original_name=Linear)
    )
  )
)

**Comment by GuanLuo** [comment]
What is the client you used to send the inference request, it seems like the client is sending batch-1 request and thus the output will also be batch-1

**Comment by llustc** [comment]
> What is the client you used to send the inference request, it seems like the client is sending batch-1 request and thus the output will also be batch-1

My batch is placed in the second dimension. When batch=1, the normal output is (81, 1, 7219). If batch=n, the output should be (81, n, 7219), but the output of triton is (1, n, 7219)

**Comment by GuanLuo** [comment]
So for your model, the input batch dimension is the first dimension while the output batch dimension is the second dimension? Triton's dynamic batching assumes the batch dimension is always the first dimension and it will not work properly in this case. Since 81 is at the first dimension of the output, I suspect Triton treats that as batch dimension and scatter 1 in the 81 for the response as the request batch size is 1. You will need to set `max_batch_size: 0` to disable dynamic batching and let your client send the whole batch as one request.

However, it appears that the PyTorch backend doesn't validate model output against the shape in model config, perhaps the check is omitted for performance reason. @CoderHam can you confirm? If that is the case, we may want additional configuration to enable the checking if desired, so the client will receive actual error instead of wrong result for debugging.

**Comment by CoderHam** [comment]
@GuanLuo yes the check for the output shape was omitted for performance reasons. We can add strict checking for the output shapes based on model config. 

Keep in mind we do not have this output validation post inference for any of the backends during inference. The reason we do not have this during model loading (like the other backends) is because pytorch does not store sufficient metadata about shape / dtype for us to validate the same. 

cc @tanmayv25 for viz

**Comment by llustc** [comment]
> So for your model, the input batch dimension is the first dimension while the output batch dimension is the second dimension? Triton's dynamic batching assumes the batch dimension is always the first dimension and it will not work properly in this case. Since 81 is at the first dimension of the output, I suspect Triton treats that as batch dimension and scatter 1 in the 81 for the response as the request batch size is 1. You will need to set `max_batch_size: 0` to disable dynamic batching and let your client send the whole batch as one request.
> 
> However, it appears that the PyTorch backend doesn't validate model output against the shape in model config, perhaps the check is omitted for performance reason. @CoderHam can you confirm? If that is the case, we may want additional configuration to enable the checking if desired, so the client will receive actual error instead of wrong result for debugging.

Thanks for your reply, I solved the problem using the way you provided.
